Refactors CRS code into a class

jameshiebert opened the pull request:
This PR replaces the coordinate reference system (crs) as a dict with a CRS class. It resolves #736 through a number of changes:

* Creates a `CRS` class that simply inherits from `UserDict`
* Moves `CRS` related methods (`to_string`, `from_string`, `is_geographic`, `is_projected`, `is_valid`, and `is_epsg_code`) to be class methods
* Moves the `CRS` comparison method to be `CRS.__eq__` (allowing comparison through the `==` operator)
* Modifies tests that compare `CRS` output to a raw dict, to compare to `UserDict.data` property
* Replaces all creation of `CRS` representation using bare dicts with the use class instantiation
* Removes the `crs_wkt` property from `DatasetReader` class

sgillies commented:
@jameshiebert I think this is good to go. I think I'd like to see `is_geographic` etc be properties instead of methods. How does that sound to you?

brendan-ward commented on rasterio/_crs.pyx:
```diff
@@ -0,0 +1,283 @@
+        return self.is_geographic() or self.is_projected()
+
+    def is_geographic(self):
+        cdef void *osr_crs = _base._osr_from_crs(self)
```
@jameshiebert I'm wondering if we should wrap these in `try` / `finally` to prevent memory leaks in case we don't deallocate `osr_crs`.

Untested:

```
cdef void *osr_crs = _base._osr_from_crs(self)
try: 
    cdef int retval = _gdal.OSRIsGeographic(osr_crs)
    return retval == 1
finally:
    _gdal.OSRDestroySpatialReference(osr_crs)
```

brendan-ward replied:
Or this could possibly be wrapped in a context manager to do that for you, for this and the following very similar lines.  I haven't done that in Cython yet, so I don't know if it is easy / possible.  Feel free to save that for a later refactor too.

jameshiebert commented:
Both of those sound like great suggestions. Thanks for the feedback.

brendan-ward commented on rasterio/rio/edit_info.py:
```diff
@@ -7,8 +7,10 @@
 from . import options
 import rasterio
-import rasterio.crs
+from rasterio._crs import CRS
```
@jameshiebert should this be `from rasterio.crs import CRS` instead?  We shouldn't be pulling the Cython interface into places other than into `crs.py`

jameshiebert replied:
Yeah, that's a good point. We should probably keep the import consistent across the board.

brendan-ward commented on rasterio/_io.pyx:
```diff
@@ -1479,6 +1479,8 @@ cdef class RasterUpdater(RasterReader):
         # Normally, we expect a CRS dict.
         if isinstance(crs, dict):
+            crs = CRS(crs)
+        if isinstance(crs, CRS):
```
@jameshiebert Should the order of 1481 and 1483 be flipped?  Since we now prefer `CRS` objects, I think we should test that first.

jameshiebert replied:
Well, the two conditionals are not mutually exclusive, since we modify the `crs` object in the first block if it comes in as a dict. Then we *also* proceed to do everything in the next block, because now it *is* an instance of `CRS`. We do prefer `CRS` object, but we're saying that if it comes in as a `dict`, make it a `CRS` object and then proceed as normal. If we flipped the conditionals, then nothing in the second conditional would get executed when a `dict` comes in, but we want it to be.

brendan-ward commented on rasterio/rio/rasterize.py:
```diff
@@ -148,7 +154,7 @@ def feature_value(feature):
         if os.path.exists(output):
             with rasterio.open(output, 'r+') as out:
-                if has_src_crs and not is_same_crs(src_crs, out.crs):
+                if has_src_crs and src_crs != out.crs:
```
👍  nice change!

brendan-ward commented on tests/test_dataset_mask.py:
```diff
@@ -53,7 +53,7 @@ def tiffs(tmpdir):
         'affine': Affine(5.0, 0.0, 0.0, 0.0, -5.0, 0.0),
         'transform': Affine(5.0, 0.0, 0.0, 0.0, -5.0, 0.0),
-        'crs': {'init': 'epsg:4326'},
+        'crs': CRS(),
```
@jameshiebert please make this explicit rather than implicit:

`'crs': CRS({'init': 'epsg:4326'})`

brendan-ward commented on tests/test_crs.py:
```diff
@@ -18,15 +16,15 @@
 def test_read_epsg(tmpdir):
     with rasterio.open('tests/data/RGB.byte.tif') as src:
-        assert src.crs == {'init': 'epsg:32618'}
+        assert src.crs.data == {'init': 'epsg:32618'}
```
I'm finding that `crs.data` is predicated on the user knowing that `CRS` overrides `UserDict`.  What about adding a new property to CRS that returns `self.data`, so that in usages like this we can:

`assert src.crs.to_dict == {'init': 'epsg:32618'}`

jameshiebert replied:
Yeah, I hadn't really intended for `crs.data` to be used as a public interface, but I guess that if I'm using it in the tests, then it is. `to_dict` sounds like a reasonable idea, though I would think it best to be a method rather than a property (to_ implies that some action is being taken).

brendan-ward replied:
@jameshiebert Yes, agreed about method instead of property.  My bad.

perrygeo commented:
@jameshiebert This really looks like a solid improvement. Subclassing from a dict and keeping the equality operator is a great move. 

Two questions:

* should we update the docstrings in [`warp.py`](https://github.com/jameshiebert/rasterio/blob/issue-736/rasterio/warp.py) to indicate that both dicts and CRS objects are accepted?

*  In an effort to move the pure python logic out of the `.pyx` files, what do think about having a `rasterio._crs._CRS` base class and defining `rasterio.crs.CRS` as a subclass where we would add the pure python methods? Might be more trouble than it's worth in this case. In general though, I find that unless you're touching C code or getting a speed boost from static typing, Cython code complicates development - compile times, lack of debugging/profiling/coverage tools, etc. This is also something that we can defer to post-1.0 since it wouldn't change the interface. Just a thought.

jameshiebert commented:
@perrygeo Good suggestions on both fronts and I pushed patches that address them. I'm of a bit mixed opinion on the splitting of `_crs.pyx` and `crs.py`. I can see the advantages, particularly if there is a pretty clear separation between the C and Python responsibilities. OTOH, it has the potential to obsfucate the code a bit and split related functionality across multiple files. But if you guys are happy with doing so, it is now implemented that way.

brendan-ward commented on rasterio/_crs.pyx:
```diff
@@ -0,0 +1,56 @@
+            _gdal.OSRDestroySpatialReference(osr_crs)
+
+    def __eq__(self, other):
+        cdef void *osr_crs1 = _base._osr_from_crs(self)
```
@jameshiebert this is another place where `try` / `finally` might be appropriate too.

For the above `try` / `finally` they are probably OK as written, but I think my preference would be (untested code):
```
cdef void *osr_crs = NULL
cdef int retval
try:
    osr_crs = _base._osr_from_crs(self)
    retval = _gdal.OSRIsGeographic(osr_crs)
    return bool(retval == 1)
finally:
    _gdal.OSRDestroySpatialReference(osr_crs)
```

Just to avoid any possibility of leaving that `osr_crs` lying around.

brendan-ward commented on rasterio/crs.py:
```diff
@@ -1,111 +1,130 @@
+
+    @property
+    def is_epsg_code(self):
+        for val in self.values():
```
@jameshiebert PEP8 isn't helping w/ readability here.  Would changing `val` to `v` let this fit on one line?

If not, how about a local variable to split this up:

```
is_str = isinstance(val, string_types)
if is_sr and val.lower().startswith('epsg'):
    return True
```

perrygeo replied:
@brendan-ward agreed. By my count the whole line would be 80 chars exactly so let's just keep it as one line. As it says in PEP8 "However, know when to be inconsistent -- sometimes style guide recommendations just aren't applicable. When in doubt, use your best judgment"

jameshiebert replied:
Looks like the newline is the 80th character, so one line sounds fine.

perrygeo commented:
@jameshiebert  Yep, the cython files encapsulate the interface with GDAL while the python files define the public interface. It's a useful split. I can see it being a little awkward but I think benefits are worth it.

The two remaining comments should be quick to address - then we are good to go 🚀 Ping me when it's ready to merge.

jameshiebert commented:
@perrygeo The last two commits should address the remaining comments. Feel free to merge if it looks good to you.
